### PR TITLE
Feature resource contribution

### PR DESCRIPTION
**Pull request by andela-uibrahim**
## What does this PR do?

Adds a user resource contributions endpoint. (/api/v1/resources) that responds when a users contributes resource on learning map.

## Description of tasks Completed
- Design the data models for suggestions
- Develop the data models for suggestions
- Write tests for the data models
- Document API endpoints on Apiary
- Create endpoint /suggestion
- Write integration test for the endpoint

## How should this be manually tested?

At the root of the project directory run the following command:

$pytest -v api/test/test_resource.py

## What are the relevant pivotal tracker stories?

#148227477

### CODE REVIEW

File: api/.env [hunk 1]
```diff
@@ -0,0 +1,3 @@
+SQLALCHEMY_DATABASE_URI_2=postgresql://freemile:Kratus043@127.0.0.1:5432/learning-map-api-test
```
Comment (njirap) [inline api/.env]: this is what was giving me problems locally.. Chai

Comment (njirap) [inline api/.env]: this is to be ignored

File: api/helpers.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+            return False
+        return True
+    def validate_list(self, input_type):
+        if type(input_type) is not list:
+            return False
+        return True
+    def validate_integer(self, input_type):
+        if type(input_type) is not int:
+            return False
+        return True
```
Comment (njirap) [inline api/helpers.py]: files should end in blank lines

File: api/test/api/test_resource.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
+import json
+
+from flask_restful import Api
+
+from learning_map_api.api.test.base import BaseTestCase
+from learning_map_api.api.models import db
+from learning_map_api.api.test.base import BaseTestCase
```
Comment (njirap) [inline api/test/api/test_resource.py]: double importing.. already L5

**Comment by njirap** [comment]
@andela-uibrahim Please remember not to push your `.env` Sir

File: api/.env (hunk 1 above)
Comment (andela-uibrahim) [inline api/.env]: haha okay, noted

**Comment by njirap** [comment]
@andela-uibrahim what plans do you have for this PR?

File: .gitignore [hunk 4]
```diff
@@ -1,10 +1,10 @@
...
 __pycache_/
 .cache/
-.env
 .DS_Store
 .idea/
 instance/
 migrations/
 .vscode/
-*.sqlite
+test_db.sqlite
```
Comment (njirap) [inline .gitignore]: the `*.sqlite` that you are removing already covers this.. please undo this change

File: .gitignore [hunk 5]
```diff
@@ -1,10 +1,10 @@
...
+.env
```
Comment (njirap) [inline .gitignore]: please remove this. if you have this `.env` locally then it can live outside the project, otherwise you are advised to use a virtual env wrapper

File: api/helpers.py (hunk 2 above)
Comment (njirap) [inline api/helpers.py]: @andela-uibrahim what is the meaning of the thumbs up if you were not going to add the blank line?

File: api/models.py [hunk 6]
```diff
@@ -39,7 +39,6 @@ class Tag(db.Model, SerializerMixin):
     """
     The Tag Model
     """
-
```
Comment (njirap) [inline api/models.py]: why have you removed this line?

File: api/models.py [hunk 7]
```diff
@@ -58,7 +57,7 @@ class Contribution(db.Model, SerializerMixin):
     created_at = db.Column(db.DateTime, default=datetime.utcnow())
     user_id = db.Column(db.String, nullable=False)
     approver_id = db.Column(db.String)
-    tags = db.relationship("Tag", secondary="contribution_tag", backref="contributions",
```
Comment (njirap) [inline api/models.py]: why have you removed the quote?

Comment (andela-uibrahim) [inline api/models.py]: am sorry I thought it was an omission since it was referencing an existing variable. that was why i removed it

File: api/models.py (hunk 6 above)
Comment (andela-uibrahim) [inline api/models.py]: adding it back rightaway

File: api/views/contribution_types.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
+import uuid
+
+from flask import request, jsonify
+from flask_restful import Resource
+
+from ..models import db, ContributionType
+
+
+class ContributionTypeResource(Resource):
```
Comment (njirap) [inline api/views/contribution_types.py]: have you consulted with our front end stakeholders to confirm if this is necessary?

Comment (andela-uibrahim) [inline api/views/contribution_types.py]: No i haven't. The field was present in the contributions model. so I felt finding a way to create it may be necessary. will get in touch with kes on this.

File: api/helpers.py [hunk 9]
```diff
@@ -0,0 +1,13 @@
+class Validation:
+    def validate_string(self, input_type):
+        if type(input_type) is not str:
+            return False
+        return True
```
Comment (njirap) [inline api/helpers.py]: conditionals that return booleans are redundant..this is because conditions themselves e.g if already return booleans.. therefore this function can be 1 line of code `return type(input_type) is str`

Comment (andela-oowonikoko) [inline api/helpers.py]: Sooo true 👍

File: api/views/contribution_types.py (hunk 8 above)
Comment (njirap) [inline api/views/contribution_types.py]: ok.. I also got in touch with them and this is not required.

File: api/helpers.py (hunk 9 above)
Comment (andela-uibrahim) [inline api/helpers.py]: wow, so cool

File: api/helpers.py [hunk 10]
```diff
@@ -0,0 +1,7 @@
+class Validation:
+    def validate_string(self, input_type):
+        return type(input_type) is str
+    def validate_list(self, input_type):
+        return type(input_type) is list
+    def validate_integer(self, input_type):
+        return type(input_type) is int
```
Comment (njirap) [inline api/helpers.py]: all files must end in blank line

Comment (njirap) [inline api/helpers.py]: how many more times do I have to say this please?

Comment (njirap) [inline api/helpers.py]: In the spirit if DRY this can all be compressed to e.g
```
def validate_type(item, input_type):
    return type(item) is input_type
```

Comment (njirap) [inline api/helpers.py]: it doesn't have to be a class it can be a function.

File: api/views/contributions.py [hunk 11]
```diff
@@ -1,15 +1,24 @@
...
+except ModuleNotFoundError:
+    from learning_map_api.api.helpers import Validation
+
+validator = Validation()
+
 
 
 class ContributionsResource(Resource):
 
     def get(self, user_id):
```
Comment (njirap) [inline api/views/contributions.py]: what's the purpose of the blank line?

File: api/views/contributions.py [hunk 12]
```diff
@@ -28,4 +37,89 @@ def get(self, user_id):
         #
         #     # append resource to contributions list
         #     contributions.append(user_resource_serialized)
-        return jsonify(dict(status=200, contributions=contributions))
+        # return jsonify(dict(status=200, contributions=contributions))
```
Comment (njirap) [inline api/views/contributions.py]: do not push commented code..

Comment (njirap) [inline api/views/contributions.py]: does this `get` function end here? It does not apper to be  responding if `_user_contributions` exists

Comment (andela-uibrahim) [inline api/views/contributions.py]: this code was inherited from develop. commented out that way

File: api/views/contributions.py [hunk 13]
```diff
@@ -28,4 +37,89 @@ def get(self, user_id):
...
+            for key in keys:
+                if key not in result:
+                    return {"message": key + " cannot be null"}, 400
+                elif (key in ["title", "description", "type"]) and not validator.validate_string(result[key]):
+                    return {"message": key + " must be a valid string"}, 400
+                elif (key is "tags") and not validator.validate_list(result[key]):
+                    return {"message": key + " must be a list"}, 400
+            contribution_types = ['resource', 'idea', 'skill']
+        
+            if result['type'].lower() not in contribution_types:
```
Comment (njirap) [inline api/views/contributions.py]: I don't envision a word where this string is the contribution type's name but the `id` please adjust accordingly

File: api/views/contributions.py [hunk 14]
```diff
@@ -28,4 +37,89 @@ def get(self, user_id):
         #
         #     # append resource to contributions list
         #     contributions.append(user_resource_serialized)
-        return jsonify(dict(status=200, contributions=contributions))
+        # return jsonify(dict(status=200, contributions=contributions))
+    
+    def post(self, user_id):
```
Comment (njirap) [inline api/views/contributions.py]: where is this `id` coming from?

1. Please don't tell me it's from the url
2. the frontend is sending us tokens to work with not user ids

File: api/helpers.py (hunk 10 above)
Comment (andela-uibrahim) [inline api/helpers.py]: wow i am loving this

File: api/views/contributions.py [hunk 15]
```diff
@@ -28,4 +37,89 @@ def get(self, user_id):
...
+                elif (key in ["title", "description", "type"]) and not validator.validate_string(result[key]):
+                    return {"message": key + " must be a valid string"}, 400
+                elif (key is "tags") and not validator.validate_list(result[key]):
+                    return {"message": key + " must be a list"}, 400
+            contribution_types = ['resource', 'idea', 'skill']
+        
+            if result['type'].lower() not in contribution_types:
+                return {"message": "type must be of resource, idea or skill"}, 400
+
+            contribution_type = ContributionType.query.filter_by(name=result["type"].lower()).first()
```
Comment (njirap) [inline api/views/contributions.py]: we agreed we work with ids and not names.. please adhere

File: api/views/contributions.py [hunk 16]
```diff
@@ -28,4 +37,89 @@ def get(self, user_id):
...
+    
+            if not contribution_type:
+                generated_id = uuid.uuid4().hex
+                contribution_type = ContributionType(name=result["type"], id=generated_id)
+                db.session.add(contribution_type)
```
Comment (njirap) [inline api/views/contributions.py]: the front end team already said this is not required. Please don't give me the additional work of repeating myself on this your PR.. If i want to sing songs I will go for karaoke

File: api/views/resource.py [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+from flask_restful import fields, Resource, Api
+from flask_marshmallow import Marshmallow
+from flask import request, jsonify
+
+from ..models import Resource as Resource_model, Tag, Link, db
+from learning_map_api.api.helpers import Validation
+
+validator = Validation()
+
+class ResourceResource(Resource):
```
Comment (njirap) [inline api/views/resource.py]: what is this doing that the Contribution resource cannot do?

File: api/views/resource.py [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+
+            # serialize resource
+            user_resource_serialized = new_resource.serialize()
+                # add extra keys to serialized resource
+            user_resource_serialized['message'] = "Resource was succesfully added"
+            user_resource_serialized['tags'] = [ tag.serialize() for tag in new_resource.tags]
+            user_resource_serialized['links'] = [ link.serialize() for link in new_resource.links]
+            response =  jsonify(dict(data=user_resource_serialized))
+            response.status_code=201
+            return response
```
Comment (njirap) [inline api/views/resource.py]: consider this files null and void and a waste of my time

File: api/views/contributions.py (hunk 16 above)
Comment (andela-uibrahim) [inline api/views/contributions.py]: I am kinda confused. since i am not to create an endpoint for the contribution types, with which the field is required in the contribution page. am I to assume the types already exist in the db? Thanks

File: config.py [hunk 19]
```diff
@@ -14,19 +14,16 @@ class Config(object):
     SQLALCHEMY_ECHO = False
     SQLALCHEMY_TRACK_MODIFICATIONS = False
 
-
```
Comment (njirap) [inline config.py]: why are these blank lines removed here?

File: api/views/resource.py (hunk 17 above)
Comment (andela-uibrahim) [inline api/views/resource.py]: i have deleted the file locally. was going to push it soon.

File: api/views/resource.py (hunk 18 above)
Comment (andela-uibrahim) [inline api/views/resource.py]: i have deleted the file locally. was going to push it soon.

File: api/views/contributions.py (hunk 12 above)
Comment (njirap) [inline api/views/contributions.py]: my friend if you can see Github has highlited it as a change made in this commit.. Either you get rid of this change or file a bug with gittelling them they assign changes to commits where they don't belong

File: api/views/contributions.py (hunk 16 above)
Comment (njirap) [inline api/views/contributions.py]: Yes Sir, the contributions types are Read-only for now.. hence pre-set in the db

File: main.py [hunk 20]
```diff
@@ -21,6 +23,7 @@
     )
 
 
+
```
Comment (njirap) [inline main.py]: too many blank lines here

File: main.py [hunk 21]
```diff
@@ -42,8 +45,9 @@ def index():
     # create endpoints
     api = Api(app)
     api.add_resource(ContributionsResource,
-                     '/api/v1/contributions',
+                     '/api/v1/contributions/<string:user_id>',
```
Comment (njirap) [inline main.py]: why do you resurrect that which The Lord has killed?

File: main.py [hunk 22]
```diff
@@ -42,8 +45,9 @@ def index():
...
                      endpoint='contributions')
+    # api.add_resource(IdeaResource, '/api/v1/idea')
```
Comment (njirap) [inline main.py]: first you add it, then you comment it out.. please remove this Sir..

File: requirements.txt [hunk 23]
```diff
@@ -1,24 +1,26 @@
...
+psycopg2
+py
+pytest
+python-dateutil
+python-dotenv
+python-editor
+pytz
+six
+SQLAlchemy
+Werkzeug
```
Comment (njirap) [inline requirements.txt]: how is this change related to your task?

Comment (andela-uibrahim) [inline requirements.txt]: this were added after rebasing develop. coming from develop

File: api/views/resource.py (hunk 17 above)
Comment (njirap) [inline api/views/resource.py]: thanks you Sir

File: requirements.txt (hunk 23 above)
Comment (njirap) [inline requirements.txt]: @andela-uibrahim We need to talk about your accusing develop of tingz..

File: api/views/contributions.py (hunk 12 above)
Comment (andela-uibrahim) [inline api/views/contributions.py]: njirap i am your friend oooo lol.

File: requirements.txt (hunk 23 above)
Comment (andela-uibrahim) [inline requirements.txt]: okay

Comment (andela-uibrahim) [inline requirements.txt]: @njirap  are you there?

Comment (njirap) [inline requirements.txt]: hey.. hehe let's talk tomorrow.. goodnight Sir..

Comment (njirap) [inline requirements.txt]: @andela-uibrahim go sleep I want you fresh tomorrow for another exchange of strong words hehehe

Comment (andela-uibrahim) [inline requirements.txt]: lol

File: api/views/contributions.py [hunk 24]
```diff
@@ -29,3 +34,80 @@ def get(self, user_id):
...
+                    return {"message": key + " must be a list"}, 400
+        
+            contribution_type = ContributionType.query.filter_by(id=result["type"]).first()           
+            if not contribution_type:
+                return {"message": "type must be of resource, idea or skill"}, 400
+        
+            contribution_type_id =  contribution_type.id
+
+            generated_id = uuid.uuid4().hex
+            result["status"] = "pending"
```
Comment (njirap) [inline api/views/contributions.py]: this is not necessary as the `Contribution` model already handles it.

File: api/views/contributions.py [hunk 25]
```diff
@@ -29,3 +34,80 @@ def get(self, user_id):
...
+            )
+            db.session.add(new_contribution)
+            db.session.commit()
+
+            contribution_tags = []
+            contribution_id = new_contribution.id
+
+            for tag in result["tags"]:
+                existing_tag = None
+                existing_tag = Tag.query.filter_by(name=tag).first()
```
Comment (njirap) [inline api/views/contributions.py]: let get these by `id`s and not `name`

File: api/views/contributions.py [hunk 26]
```diff
@@ -29,3 +34,80 @@ def get(self, user_id):
...
+                    contribution_tags.append(existing_tag)
+                else:
+                    tag_generated_id = uuid.uuid4().hex
+                    tag = Tag(
+                        id=tag_generated_id,
+                        name=tag
+                    )
+                    db.session.add(tag)
+                    db.session.commit()
+                    contribution_tags.append(tag)
```
Comment (njirap) [inline api/views/contributions.py]: lets not worry ourselves with this for now.

File: api/views/contributions.py [hunk 27]
```diff
@@ -29,3 +34,80 @@ def get(self, user_id):
...
+
+            for contribution_tag in contribution_tags:
+                previously_tagged = None
+
+                if contribution_tag in new_contribution.tags:
+                    previously_tagged = True
+
+                if not previously_tagged:
+                    new_contribution.tags.append(contribution_tag)
```
Comment (njirap) [inline api/views/contributions.py]: this can easily be handled from the `contribution_tag` model side

File: api/views/contributions.py [hunk 28]
```diff
@@ -29,3 +34,80 @@ def get(self, user_id):
...
+
+            contribution_serialized = new_contribution.serialize()
+            contribution_serialized["success"] = "Contribution was succesfully created"
```
Comment (njirap) [inline api/views/contributions.py]: this field misleads to make us think there's a field called success

File: test/api/views/test_contributions.py [hunk 29]
```diff
@@ -1,9 +1,273 @@
...
+
+        self.contribution_type_1 = ContributionType(
+            id = "-KA3Q0D45T6",
+            name="resource",
+        )
+        self.contribution_type_2 = ContributionType(
+            id = "-KQ0X2FG8V",
+            name="idea",
+        )
+        self.contribution_type_3 = ContributionType(
```
Comment (njirap) [inline test/api/views/test_contributions.py]: better var names are possible here is self.idea, self.skill and self.resource

File: test/api/views/test_contributions.py [hunk 30]
```diff
@@ -1,9 +1,273 @@
...
+                "user_id": "user_id"
+            }
+        }
+        generated_response = json.loads(res.data)
+        response_keys = [ "user_id", "approver_id", "status", "title", "description", "success"]
+       
+        self.assertEqual(res.status_code, 201)
+        self.assertIn("Contribution was succesfully created", str(res.data))
+        for response_key in response_keys:
+            self.assertEqual(generated_response['data'][response_key], expected_response['data'][response_key])
```
Comment (njirap) [inline test/api/views/test_contributions.py]: this is not necessary querying the database and asserting the count has incremented by 1 is sufficient

File: test/api/views/test_contributions.py [hunk 31]
```diff
@@ -1,9 +1,273 @@
...
+            
+
```
Comment (njirap) [inline test/api/views/test_contributions.py]: too many blank lines here.. supposed to be only 1

File: test/api/views/test_contributions.py [hunk 32]
```diff
@@ -1,9 +1,273 @@
...
+        res = self.client.post('/api/v1/contributions', data=json.dumps({
+            "title": "usman",
+            "type": "-KA3Q0D45T6",
+            "description": "usman baba",
+            "tags": ["usman21e062", "seune20611", "map1e2061"]
+        }), content_type="Text(text/plain)")
+
+        expected_response = {
+            "message": "Request must be a valid JSON"
+        }
```
Comment (njirap) [inline test/api/views/test_contributions.py]: this doesn't have to be a dictionary it can just be the error message string

File: test/api/views/test_contributions.py [hunk 33]
```diff
@@ -1,9 +1,273 @@
...
+        """
+
+        res = self.client.post('/api/v1/contributions', data=json.dumps({
+            "title": "usman",
+            "type": "-KA3Q0D45T6",
+            "description": "usman baba",
+        }), content_type="application/json")
+
+        expected_response = {
+            "message": "tags cannot be null"
```
Comment (njirap) [inline test/api/views/test_contributions.py]: why can't tags be null?

File: test/api/views/test_contributions.py [hunk 34]
```diff
@@ -1,9 +1,273 @@
...
+
+        res = self.client.post('/api/v1/contributions', data=json.dumps({
+            "title": 2,
+            "type": "-KA3Q0D45T6",
+            "description": "usman baba",
+            "tags": ["usman21e062", "seune20611", "map1e2061"]
+        }), content_type="application/json")
+
+        expected_response = {
+            "message": "title must be a valid string"
```
Comment (njirap) [inline test/api/views/test_contributions.py]: cover this with the other error below

File: test/api/views/test_contributions.py [hunk 35]
```diff
@@ -1,9 +1,273 @@
...
+
+        res = self.client.post('/api/v1/contributions', data=json.dumps({
+            "title": "usman",
+            "type": 4,
+            "description": "usman baba",
+            "tags": ["usman21e062", "seune20611", "map1e2061"]
+        }), content_type="application/json")
+
+        expected_response = {
+            "message": "type must be a valid string"
```
Comment (njirap) [inline test/api/views/test_contributions.py]: must be a valid type_id
